relayer/chains/substrate/query.go

```diff
@@ -0,0 +1,519 @@
+package substrate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	transfertypes "github.com/cosmos/ibc-go/v5/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
+	conntypes "github.com/cosmos/ibc-go/v5/modules/core/03-connection/types"
+	chantypes "github.com/cosmos/ibc-go/v5/modules/core/04-channel/types"
+	committypes "github.com/cosmos/ibc-go/v5/modules/core/23-commitment/types"
+	ibcexported "github.com/cosmos/ibc-go/v5/modules/core/exported"
+	"github.com/cosmos/relayer/v2/relayer/provider"
+	"golang.org/x/sync/errgroup"
+)
+
+var _ provider.QueryProvider = &SubstrateProvider{}
+
+// QueryTx takes a transaction hash and returns the transaction
+func (sp *SubstrateProvider) QueryTx(ctx context.Context, hashHex string) (*provider.RelayerTxResponse, error) {
+	return &provider.RelayerTxResponse{}, errors.New(ErrTextSubstrateDoesNotHaveQueryForTransactions)
+}
+
+// QueryTxs returns an array of transactions given a tag
+func (sp *SubstrateProvider) QueryTxs(ctx context.Context, page, limit int, events []string) ([]*provider.RelayerTxResponse, error) {
+	return []*provider.RelayerTxResponse{}, errors.New(ErrTextSubstrateDoesNotHaveQueryForTransactions)
+}
+
+// QueryBalance returns the amount of coins in the relayer account
+func (sp *SubstrateProvider) QueryBalance(ctx context.Context, keyName string) (sdk.Coins, error) {
+	var (
+		addr string
+		err  error
+	)
+	if keyName == "" {
+		addr, err = sp.Address()
+	} else {
+		sp.Config.Key = keyName
```

Comment:
This would change the SubstrateProvider's configured key in memory but not on disk so later operations could be referencing two different keys, the one on disk in the config file or the one now referenced in memory. As a consumer of this API it would be an unexpected side effect for a method named `QueryBalance` to modify state instead of just querying some data. In the cosmos implementation we just throw an error if the `keyName` argument is invalid for any reason whether it's an empty string or there is no key that matches the name.

Reply:
Yeah, that makes sense. I'll resolve it in another PR.